Parsedown.php

```diff
@@ -1485,10 +1503,80 @@ function parse($text)
         return $markup;
     }
 
+    protected function sanitiseElement(array $Element)
+    {
+        static $badAttributeChars = "\"'= \t\n\r\0\x0B";
+        static $safeUrlNameToAtt  = array(
+            'a'   => 'href',
+            'img' => 'src',
+        );
+
+        if (isset($safeUrlNameToAtt[$Element['name']]))
+        {
+            $Element = $this->filterUnsafeUrlInAttribute($Element, $safeUrlNameToAtt[$Element['name']]);
+        }
+
+        if ( ! empty($Element['attributes']))
+        {
+            foreach ($Element['attributes'] as $att => $val)
+            {
+                # clear out nulls
+                if ($val === null)
+                {
+                    unset($Element['attributes'][$att]);
+                }
+                # filter out badly parsed attribute
+                elseif (strpbrk($att, $badAttributeChars) !== false)
+                {
+                    unset($Element['attributes'][$att]);
+                }
+            }
+
+            $onEventAttributeKeys = preg_grep('/^on/i', array_keys($Element['attributes']));
+
+            foreach ($onEventAttributeKeys as $att)
+            {
+                unset($Element['attributes'][$att]);
+            }
+        }
+
+        return $Element;
+    }
+
+    protected function filterUnsafeUrlInAttribute(array $Element, $attribute)
+    {
+        if ($this->safeLinksEnabled)
+        {
+            $safe = false;
+
+            foreach ($this->safeLinksWhitelist as $scheme)
+            {
+                if (stripos($Element['attributes'][$attribute], $scheme) === 0)
```

Comment:
> I don't think that re-inventing the wheel makes sense here. How does Symfony encode URLs? Or Zend? Or end-user software that accepts arbitrary URLs like phpBB, MediaWiki, WordPress? The list is basically endless.

A lot of software that accepts input from (untrusted) users probably won't have any reason to support relative URLs

e.g. phpBB doesn't support them 🤷‍♂️ 
<img width="623" alt="screen shot 2017-05-04 at 00 33 56" src="https://cloud.githubusercontent.com/assets/3288888/25685861/21ae1248-3062-11e7-95e7-29de6d944535.png">

---


> I'm in contrast talking about whitelisting particular protocols like http:// and ftp://, as this is, by definition, something user's will stumble upon (because there's nothing bad about irc://, steam:// and so on - the list is endless, that's problem of trying to whitelist them)

Unfortunately there's probably an endless list of ways to specify what might be useful to henceforth call a 'dirty' protocol (that isn't technically valid, but browsers will accept).
And telling the difference between a dirty protocol and a relative URL is non trivial.

I haven't worked with Symphony's internals before, so tracking down exactly how links are built in full would probably take a while for me to do.
Safe to say they do not do it by searching for the string `javascript` though 😉 

<img width="1789" alt="screen shot 2017-05-04 at 00 51 58" src="https://cloud.githubusercontent.com/assets/3288888/25686164/1c7c2bdc-3064-11e7-8b0b-771d289497dd.png">